deepgram/clients/speak/v1/async_client_stream.py

```diff
@@ -0,0 +1,579 @@
+                SpeakStreamEvents(SpeakStreamEvents.Open),
+                OpenResponse(type=SpeakStreamEvents.Open),
+            )
+
+            self._logger.notice("start succeeded")
+            self._logger.debug("AsyncSpeakStreamClient.start LEAVE")
+            return True
+        except websockets.ConnectionClosed as e:
+            self._logger.error("ConnectionClosed in AsyncSpeakStreamClient.start: %s", e)
+            self._logger.debug("AsyncSpeakStreamClient.start LEAVE")
+            if self._config.options.get("termination_exception_connect") == "true":
+                raise
+            return False
+        except websockets.exceptions.WebSocketException as e:
+            self._logger.error("WebSocketException in AsyncSpeakStreamClient.start: %s", e)
+            self._logger.debug("AsyncSpeakStreamClient.start LEAVE")
+            if self._config.options.get("termination_exception_connect") == "true":
+                raise
+            return False
+        except Exception as e:  # pylint: disable=broad-except
+            self._logger.error("WebSocketException in AsyncSpeakStreamClient.start: %s", e)
+            self._logger.debug("AsyncSpeakStreamClient.start LEAVE")
+            if self._config.options.get("termination_exception_connect") == "true":
+                raise
+            return False
+
+    # pylint: enable=too-many-branches,too-many-statements
+
+    def on(self, event: SpeakStreamEvents, handler) -> None:
+        """
+        Registers event handlers for specific events.
+        """
+        self._logger.info("event subscribed: %s", event)
+        if event in SpeakStreamEvents.__members__.values() and callable(handler):
+            self._event_handlers[event].append(handler)
+
+    # triggers the registered event handlers for a specific event
+    async def _emit(self, event: SpeakStreamEvents, *args, **kwargs) -> None:
+        """
+        Emits events to the registered event handlers.
+        """
+        self._logger.debug("AsyncSpeakStreamClient._emit ENTER")
+        self._logger.debug("callback handlers for: %s", event)
+
+        # debug the threads
+        for thread in threading.enumerate():
+            self._logger.debug("after running thread: %s", thread.name)
+        self._logger.debug("number of active threads: %s", threading.active_count())
+
+        tasks = []
+        for handler in self._event_handlers[event]:
+            tasks.append(asyncio.create_task(handler(self, *args, **kwargs)))
+
+        if len(tasks) > 0:
+            self._logger.debug("waiting for tasks to finish...")
+            await asyncio.gather(*filter(None, tasks), return_exceptions=True)
+
+        # debug the threads
+        for thread in threading.enumerate():
+            self._logger.debug("after running thread: %s", thread.name)
+        self._logger.debug("number of active threads: %s", threading.active_count())
+
+        self._logger.debug("AsyncSpeakStreamClient._emit LEAVE")
+
```

Comment:
@SandraRodgers  Can you port this over to the STT websocket?

Reply:
How does coderabbitai suggest I do this:

The _emit method effectively handles asynchronous event notifications. It uses a list comprehension to create tasks for each handler, which is efficient. However, ensure that the tasks are appropriately managed to avoid memory leaks or unawaited tasks.

Reply:
@coderabbitai

How do you suggest I do this:

The _emit method effectively handles asynchronous event notifications. It uses a list comprehension to create tasks for each handler, which is efficient. However, ensure that the tasks are appropriately managed to avoid memory leaks or unawaited tasks.